src/cascade/model/model_writer.py

```diff
@@ -0,0 +1,356 @@
+"""
+Writes a Model to a Dismod File.
+"""
+from math import nan, inf
+from numbers import Real
+
+import numpy as np
+import pandas as pd
+
+from cascade.core import getLoggers
+from cascade.dismod.constants import DensityEnum, WeightEnum, MulCovEnum
+
+CODELOG, MATHLOG = getLoggers(__name__)
+
+
+class ModelWriter:
+    """This layer over the Dismod File is responsible for translation between
+    absolute numbers in the Model and SQL-indexed numbers.
+
+    This has an API that the client has to follow. Functions need to be called
+    in order.
+
+    For naming in the DismodFile, this chooses names based on the random
+    fields.
+
+    Assumptions this ModelWriter makes, on top of what Dismod-AT requires:
+
+     * Locations have a ``c_location_id`` member that ``maps node_id``
+       to ``location_id``.
+     * Rates and integrands are always in the same order.
+    """
+
+    def __init__(self, session, extremal_age_time):
+        """
+
+        Args:
+            session (Session): The Dismod-AT Session into which this writes.
+            extremal_age_time (Tuple(Set(int),Set(int))): a set of ages and
+                a set of times to add to the age and time tables. Dismod-AT
+                requires that all data points in the avgint and data tables
+                be within the min and max of the age and time tables, so this
+                passes in the extremal values.
+        """
+        self._session = session
+        self._dismod_file = session.dismod_file
+        if extremal_age_time:
+            self._ages = np.array(list(extremal_age_time[0]), dtype=np.float)
+            self._times = np.array(list(extremal_age_time[1]), dtype=np.float)
+        else:
+            self._ages = np.zeros((0,), dtype=np.float)
+            self._times = np.zeros((0,), dtype=np.float)
```

Comment:
Do you meant `np.empty`? I'm not sure which is weirder, an array of length zero initialized with zeros (even though that means nothing since there are no elements) or an array of length zero uninitialized (even though that means nothing since there are no elements). I think I like `np.empty` better?

Reply:
Sure. This is to make append make more sense.